kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,295 @@
+			!(b >= '0' && b <= '9') && b != '-' {
+
+			return true
+		}
+	}
+
+	return false
+}
+
+func (tx *readWriteTx) toTableName(key []byte) string {
+	// Max table name length in postgres is 63, but keep some slack for
+	// index postfixes.
+	const maxTableNameLen = 50
+
+	var table string
+	if hasSpecialChars(string(key)) {
+		return ""
+	}
+
+	table = strings.Replace(string(key), "-", "_", -1)
+	table = tx.db.getPrefixedTableName(table)
+
+	if len(table) > maxTableNameLen {
+		return ""
+	}
+
+	return table
+}
+
+// ReadBucket opens the root bucket for read only access.  If the bucket
+// described by the key does not exist, nil is returned.
+func (tx *readWriteTx) ReadBucket(key []byte) walletdb.ReadBucket {
+	bucket := tx.ReadWriteBucket(key)
+	if bucket == nil {
+		return nil
+	}
+	return bucket.(walletdb.ReadBucket)
+}
+
+// ForEachBucket iterates through all top level buckets.
+func (tx *readWriteTx) ForEachBucket(fn func(key []byte) error) error {
+	// Fetch human-readable top level buckets.
+	rows, err := tx.tx.QueryContext(
+		context.TODO(),
+		"SELECT tablename FROM pg_catalog.pg_tables WHERE schemaname='public'",
+	)
+	if err != nil {
+		return err
+	}
+
+	var keys [][]byte
+	for rows.Next() {
+		var table string
+		err := rows.Scan(&table)
+		if err != nil {
+			return err
+		}
+
+		if !strings.HasPrefix(table, tx.db.prefix+"_") {
+			continue
+		}
+
+		keys = append(keys, []byte(table[len(tx.db.prefix)+1:]))
+	}
+
+	// Fetch binary top level buckets..
+	bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+	err = bucket.ForEach(func(k, v []byte) error {
+		keys = append(keys, k)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Sort keys.
+	sort.Slice(keys, func(i, j int) bool { return bytes.Compare(keys[i], keys[j]) < 0 })
+
+	// Call callback.
+	for _, k := range keys {
+		err = fn(k)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Rollback closes the transaction, discarding changes (if any) if the
+// database was modified by a write transaction.
+func (tx *readWriteTx) Rollback() error {
+	// If the transaction has been closed roolback will fail.
+	if !tx.active {
+		return walletdb.ErrTxClosed
+	}
+
+	err := tx.tx.Rollback()
+	if err != nil {
+		return err
+	}
+	tx.active = false
+
+	if tx.readOnly {
+		tx.db.lock.RUnlock()
+	} else {
+		tx.db.lock.Unlock()
+	}
+
+	return nil
+}
+
+// ReadWriteBucket opens the root bucket for read/write access.  If the
+// bucket described by the key does not exist, nil is returned.
+func (tx *readWriteTx) ReadWriteBucket(key []byte) walletdb.ReadWriteBucket {
+	table := tx.toTableName(key)
+
+	// If the key can't be mapped to a table name, open the bucket from the
+	// hex table.
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.NestedReadWriteBucket(key)
+	}
+
+	var value int
+	err := tx.tx.QueryRowContext(
+		context.TODO(),
+		"SELECT 1 FROM pg_catalog.pg_tables WHERE schemaname='public' and tablename=$1",
```

Comment:
I think it depends on what privileges the user is granted. We could create our own table with table names if we don't want to rely on the system table.

Reply:
Discussed offline, decided to write a setup doc for postgres describing permission etc. Also check out sqlite and see if it can iterate tables. If not, a dedicated table-table might be a good idea.

Reply:
For sqlite, it also seems to be possible: https://database.guide/2-ways-to-list-tables-in-sqlite-database/

Reply:
`docs/postgres.md` added.